deltatocumulativeprocessor segfaults at stream limit

### Component(s)

processor/deltatocumulative

### What happened?

## Description

When hitting its stream limit, the deltatocumulativeprocessor segfaults

## Steps to Reproduce

- Set a stream limit (e.g. 10000)
- Have more non-stale streams than limit, so nothing evicts
- Panic

## Expected Result

Excess samples are dropped and recorded as metric

## Actual Result

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xac75c4]

goroutine 7 [running]:
go.opentelemetry.io/collector/pdata/pmetric.NumberDataPoint.Attributes(...)
	/go/pkg/mod/go.opentelemetry.io/collector/pdata@v1.8.0/pmetric/generated_numberdatapoint.go:51
go.opentelemetry.io/collector/pdata/pmetric.NumberDataPoint.CopyTo({0x0?, 0x0?}, {0xc004a74360?, 0xc004618bd8?})
	/go/pkg/mod/go.opentelemetry.io/collector/pdata@v1.8.0/pmetric/generated_numberdatapoint.go:133 +0x44
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/data.Number.CopyTo({{0x0?, 0x0?}}, {{0xc004a74360?, 0xc004618bd8?}})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/data/data.go:37 +0x27
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams.Aggregate[...].func1({{0xc004a74360?, 0xc004618bd8?}})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/streams/data.go:41 +0x198
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams.Samples[...].func1()
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/streams/data.go:22 +0x16e
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams.Aggregate[...]({0x1436f00, 0xc003bad110}, {0x142cae0, 0xc000392a90})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/streams/data.go:35 +0xce
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor.(*Processor).ConsumeMetrics.func1({{0xc000d8abb8, 0xc004618bd8}, {0xc000190070, 0xc004618bd8}, {0xc004a74300, 0xc004618bd8}})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/processor.go:141 +0x205
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor.(*Processor).ConsumeMetrics.Each.func2({{0xc000d8abb8, 0xc004618bd8}, {0xc000190070, 0xc004618bd8}, {0xc004a74300, 0xc004618bd8}})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/metrics/util.go:22 +0x43
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/metrics.Filter.func1.1.1({0xc004a74300?, 0xc004618bd8?})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/metrics/util.go:12 +0x85
go.opentelemetry.io/collector/pdata/pmetric.MetricSlice.RemoveIf({0xc0001900b0?, 0xc004618bd8?}, 0xc0019e5ba8)
	/go/pkg/mod/go.opentelemetry.io/collector/pdata@v1.8.0/pmetric/generated_metricslice.go:111 +0x75
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/metrics.Filter.func1.1({0xc000190070?, 0xc004618bd8?})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/metrics/util.go:11 +0x6e
go.opentelemetry.io/collector/pdata/pmetric.ScopeMetricsSlice.RemoveIf({0xc000d8abd8?, 0xc004618bd8?}, 0xc0019e5c40)
	/go/pkg/mod/go.opentelemetry.io/collector/pdata@v1.8.0/pmetric/generated_scopemetricsslice.go:111 +0x75
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/metrics.Filter.func1({0xc000d8aba0?, 0xc004618bd8?})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/metrics/util.go:10 +0x56
go.opentelemetry.io/collector/pdata/pmetric.ResourceMetricsSlice.RemoveIf({0xc0035018f0?, 0xc004618bd8?}, 0xc0019e5cc8)
	/go/pkg/mod/go.opentelemetry.io/collector/pdata@v1.8.0/pmetric/generated_resourcemetricsslice.go:111 +0x75
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/metrics.Filter({0xc0035018f0?, 0xc004618bd8?}, 0x5?)
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/metrics/util.go:9 +0x3b
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/metrics.Each(...)
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/internal/metrics/util.go:21
github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor.(*Processor).ConsumeMetrics(0xc0003ae700, {0x143bf10, 0xc000387ef0}, {0xc0035018f0?, 0xc004618bd8?})
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor@v0.101.0/processor.go:136 +0x159
go.opentelemetry.io/collector/processor/batchprocessor.(*batchMetrics).export(0xc00039e8c0, {0x143bf10, 0xc000387ef0}, 0x2?, 0x0)
	/go/pkg/mod/go.opentelemetry.io/collector/processor/batchprocessor@v0.101.0/batch_processor.go:449 +0x175
go.opentelemetry.io/collector/processor/batchprocessor.(*shard).sendItems(0xc00039e900, 0x2)
	/go/pkg/mod/go.opentelemetry.io/collector/processor/batchprocessor@v0.101.0/batch_processor.go:261 +0x56
go.opentelemetry.io/collector/processor/batchprocessor.(*shard).processItem(0xc00039e900, {0x121b2e0?, 0xc00105fa50?})
	/go/pkg/mod/go.opentelemetry.io/collector/processor/batchprocessor@v0.101.0/batch_processor.go:235 +0x45
go.opentelemetry.io/collector/processor/batchprocessor.(*shard).startLoop(0xc00039e900)
	/go/pkg/mod/go.opentelemetry.io/collector/processor/batchprocessor@v0.101.0/batch_processor.go:220 +0x16d
created by go.opentelemetry.io/collector/processor/batchprocessor.(*shard).start in goroutine 1
	/go/pkg/mod/go.opentelemetry.io/collector/processor/batchprocessor@v0.101.0/batch_processor.go:184 +0x65
```


### Collector version

821c864db937ab7170e6c61bc096ead9cf405311

### Environment information

_No response_

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

--- a/.chloggen/deltatocumulative-drop-at-limit.yaml
+++ b/.chloggen/deltatocumulative-drop-at-limit.yaml
@@ -1,0 +1,30 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: deltatocumulative
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: properly drop samples when at limit
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [33285]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+  fixes a segfault in the limiting behavior, where streams exceeding the limit still had their samples processed.
+  due to not being tracked, this led to a nil-pointer deref
+
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: [user]

--- a/processor/deltatocumulativeprocessor/internal/metrics/data.go
+++ b/processor/deltatocumulativeprocessor/internal/metrics/data.go
@@ -4,6 +4,8 @@
 package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/metrics"
 
 import (
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
 	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/data"
 )
 
@@ -28,6 +30,12 @@ func (s Sum) Ident() Ident {
 	return (*Metric)(&s).Ident()
 }
 
+func (s Sum) Filter(expr func(data.Number) bool) {
+	s.Sum().DataPoints().RemoveIf(func(dp pmetric.NumberDataPoint) bool {
+		return !expr(data.Number{NumberDataPoint: dp})
+	})
+}
+
 type Histogram Metric
 
 func (s Histogram) At(i int) data.Histogram {
@@ -43,6 +51,12 @@ func (s Histogram) Ident() Ident {
 	return (*Metric)(&s).Ident()
 }
 
+func (s Histogram) Filter(expr func(data.Histogram) bool) {
+	s.Histogram().DataPoints().RemoveIf(func(dp pmetric.HistogramDataPoint) bool {
+		return !expr(data.Histogram{HistogramDataPoint: dp})
+	})
+}
+
 type ExpHistogram Metric
 
 func (s ExpHistogram) At(i int) data.ExpHistogram {
@@ -56,4 +70,10 @@ func (s ExpHistogram) Len() int {
 
 func (s ExpHistogram) Ident() Ident {
 	return (*Metric)(&s).Ident()
+}
+
+func (s ExpHistogram) Filter(expr func(data.ExpHistogram) bool) {
+	s.ExponentialHistogram().DataPoints().RemoveIf(func(dp pmetric.ExponentialHistogramDataPoint) bool {
+		return !expr(data.ExpHistogram{DataPoint: dp})
+	})
 }

--- a/processor/deltatocumulativeprocessor/internal/streams/data.go
+++ b/processor/deltatocumulativeprocessor/internal/streams/data.go
@@ -27,20 +27,34 @@ func Samples[D data.Point[D]](m metrics.Data[D]) Seq[D] {
 	}
 }
 
-// Aggregate each point and replace it by the result
-func Aggregate[D data.Point[D]](m metrics.Data[D], aggr Aggregator[D]) error {
+type filterable[D data.Point[D]] interface {
+	metrics.Data[D]
+	Filter(func(D) bool)
+}
+
+// Apply does dps[i] = fn(dps[i]) for each item in dps.
+// If fn returns [streams.Drop], the datapoint is removed from dps instead.
+// If fn returns another error, the datapoint is also removed and the error returned eventually
+func Apply[P data.Point[P], List filterable[P]](dps List, fn func(Ident, P) (P, error)) error {
 	var errs error
 
-	// for id, dp := range Samples(m)
-	Samples(m)(func(id Ident, dp D) bool {
-		next, err := aggr.Aggregate(id, dp)
+	mid := dps.Ident()
+	dps.Filter(func(dp P) bool {
+		id := identity.OfStream(mid, dp)
+		next, err := fn(id, dp)
 		if err != nil {
-			errs = errors.Join(errs, Error(id, err))
-			return true
+			if !errors.Is(err, Drop) {
+				errs = errors.Join(errs, err)
+			}
+			return false
 		}
+
 		next.CopyTo(dp)
 		return true
 	})
 
 	return errs
 }
+
+// Drop signals the current item (stream or datapoint) is to be dropped
+var Drop = errors.New("stream dropped") //nolint:revive // Drop is a good name for a signal, see fs.SkipAll

--- a/processor/deltatocumulativeprocessor/internal/streams/data_test.go
+++ b/processor/deltatocumulativeprocessor/internal/streams/data_test.go
@@ -19,12 +19,12 @@ import (
 var rdp data.Number
 var rid streams.Ident
 
-func BenchmarkApply(b *testing.B) {
+func BenchmarkSamples(b *testing.B) {
 	b.Run("iterfn", func(b *testing.B) {
 		dps := generate(b.N)
 		b.ResetTimer()
 
-		streams.Samples[data.Number](dps)(func(id streams.Ident, dp data.Number) bool {
+		streams.Samples(dps)(func(id streams.Ident, dp data.Number) bool {
 			rdp = dp
 			rid = id
 			return true

--- a/processor/deltatocumulativeprocessor/internal/streams/data_test.go
+++ b/processor/deltatocumulativeprocessor/internal/streams/data_test.go
@@ -4,6 +4,7 @@
 package streams_test
 
 import (
+	"math/rand"
 	"testing"
 
 	"github.com/stretchr/testify/require"
@@ -18,7 +19,7 @@ import (
 var rdp data.Number
 var rid streams.Ident
 
-func BenchmarkSamples(b *testing.B) {
+func BenchmarkApply(b *testing.B) {
 	b.Run("iterfn", func(b *testing.B) {
 		dps := generate(b.N)
 		b.ResetTimer()
@@ -55,22 +56,6 @@ func BenchmarkSamples(b *testing.B) {
 	})
 }
 
-func TestSample(t *testing.T) {
-	const total = 1000
-	dps := generate(total)
-
-	// check that all samples are visited
-	seen := 0
-	streams.Samples[data.Number](dps)(func(id streams.Ident, dp data.Number) bool {
-		require.Equal(t, dps.id, id)
-		require.Equal(t, dps.dps[seen], dp)
-		seen++
-		return true
-	})
-
-	require.Equal(t, total, seen)
-}
-
 func TestAggregate(t *testing.T) {
 	const total = 1000
 	dps := generate(total)
@@ -82,7 +67,7 @@ func TestAggregate(t *testing.T) {
 		return dp, nil
 	})
 
-	err := streams.Aggregate(dps, inv)
+	err := streams.Apply(dps, inv.Aggregate)
 	require.NoError(t, err)
 
 	// check that all samples are inverted
@@ -91,7 +76,26 @@ func TestAggregate(t *testing.T) {
 	}
 }
 
-func generate(n int) Data {
+func TestDrop(t *testing.T) {
+	const total = 1000
+	dps := generate(total)
+
+	var want []data.Number
+	maybe := aggr(func(_ streams.Ident, dp data.Number) (data.Number, error) {
+		if rand.Intn(2) == 1 {
+			want = append(want, dp)
+			return dp, nil
+		}
+		return dp, streams.Drop
+	})
+
+	err := streams.Apply(dps, maybe.Aggregate)
+	require.NoError(t, err)
+
+	require.Equal(t, want, dps.dps)
+}
+
+func generate(n int) *Data {
 	id, ndp := random.Sum().Stream()
 	dps := Data{id: id, dps: make([]data.Number, n)}
 	for i := range dps.dps {
@@ -99,7 +103,7 @@ func generate(n int) Data {
 		dp.SetIntValue(int64(i))
 		dps.dps[i] = dp
 	}
-	return dps
+	return &dps
 }
 
 type Data struct {
@@ -117,6 +121,16 @@ func (l Data) Len() int {
 
 func (l Data) Ident() metrics.Ident {
 	return l.id.Metric()
+}
+
+func (l *Data) Filter(expr func(data.Number) bool) {
+	var next []data.Number
+	for _, dp := range l.dps {
+		if expr(dp) {
+			next = append(next, dp)
+		}
+	}
+	l.dps = next
 }
 
 type aggr func(streams.Ident, data.Number) (data.Number, error)

--- a/processor/deltatocumulativeprocessor/internal/telemetry/faults_test.go
+++ b/processor/deltatocumulativeprocessor/internal/telemetry/faults_test.go
@@ -29,6 +29,7 @@ func TestFaults(t *testing.T) {
 		Pre  func(Map, identity.Stream, data.Number) error
 		Bad  func(Map, identity.Stream, data.Number) error
 		Err  error
+		Want error
 	}
 
 	sum := random.Sum()
@@ -87,7 +88,8 @@ func TestFaults(t *testing.T) {
 				dp.SetTimestamp(ts(20))
 				return dps.Store(id, dp)
 			},
-			Err: streams.ErrLimit(1),
+			Err:  streams.ErrLimit(1),
+			Want: streams.Drop, // we can't ignore being at limit, we need to drop the entire stream for this request
 		},
 		{
 			Name: "evict",
@@ -130,7 +132,7 @@ func TestFaults(t *testing.T) {
 			require.Equal(t, c.Err, err)
 
 			err = c.Bad(onf, id, dp.Clone())
-			require.NoError(t, err)
+			require.Equal(t, c.Want, err)
 		})
 	}
 }

--- a/processor/deltatocumulativeprocessor/internal/telemetry/metrics.go
+++ b/processor/deltatocumulativeprocessor/internal/telemetry/metrics.go
@@ -174,6 +174,8 @@ func (f Faults[T]) Store(id streams.Ident, v T) error {
 		inc(f.dps.dropped, reason("out-of-order"))
 	case errors.As(err, &limit):
 		inc(f.dps.dropped, reason("stream-limit"))
+		// no space to store stream, drop it instead of failing silently
+		return streams.Drop
 	case errors.As(err, &evict):
 		inc(f.streams.evicted)
 	case errors.As(err, &gap):

--- a/processor/deltatocumulativeprocessor/processor.go
+++ b/processor/deltatocumulativeprocessor/processor.go
@@ -138,7 +138,7 @@ func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) erro
 		case pmetric.MetricTypeSum:
 			sum := m.Sum()
 			if sum.AggregationTemporality() == pmetric.AggregationTemporalityDelta {
-				err := streams.Aggregate(metrics.Sum(m), p.sums.aggr)
+				err := streams.Apply(metrics.Sum(m), p.sums.aggr.Aggregate)
 				errs = errors.Join(errs, err)
 				sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
 			}
@@ -147,7 +147,7 @@ func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) erro
 		case pmetric.MetricTypeExponentialHistogram:
 			expo := m.ExponentialHistogram()
 			if expo.AggregationTemporality() == pmetric.AggregationTemporalityDelta {
-				err := streams.Aggregate(metrics.ExpHistogram(m), p.expo.aggr)
+				err := streams.Apply(metrics.ExpHistogram(m), p.expo.aggr.Aggregate)
 				errs = errors.Join(errs, err)
 				expo.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
 			}

